EtherKit/Models/TypedData.swift

```diff
@@ -0,0 +1,394 @@
+      return data
+    case let .string(value):
+      let stringData = Data(bytes: Array(value.utf8))
+      return stringData.sha3(.keccak256)
+    case let .bytes(count, value):
+      if count == UnformattedDataMode.unlimited {
+        return value.sha3(.keccak256)
+      } else {
+        var data = Data()
+        let padding = 32 - value.count
+        data.append(value)
+        data.append(Data(repeating: 0, count: padding))
+        return data
+      }
+    case let .array(_, value):
+      var data = Data()
+      for subVal in value {
+        let subData = try subVal.encodeData(types: types)
+        data.append(subData)
+      }
+      return data
+    case let .object(type, value):
+      let typeHash = try types.getTypeHash(for: type)
+      let encodedData = try value.encodeData(typeName: type, types: types)
+      return (typeHash + encodedData).sha3(.keccak256)
+    }
+  }
+}
+
+public struct MessageObject {
+  public var values: [String: MessageValue] = [:]
+
+  public init(from object: [String: Any], typeName: String, types: Types) throws {
+    guard let type = types[typeName] else {
+      throw EtherKitError.web3Failure(reason: .parsingFailure)
+    }
+    for prop in type.properties {
+      guard let valueObject: Any = object[prop.name] else {
+        throw EtherKitError.web3Failure(reason: .parsingFailure)
+      }
+      values[prop.name] = try MessageValue.getMessageValue(from: valueObject, type: prop.type, types: types)
+    }
+  }
+
+  public func encodeData(typeName: String, types: Types) throws -> Data {
+    guard let mainType = types[typeName] else {
+      fatalError()
+    }
+
+    var data = Data()
+    for prop in mainType.properties {
+      guard let subVal = values[prop.name] else {
+        fatalError()
+      }
+      guard let subData = try? subVal.encodeData(types: types) else {
+        fatalError()
+      }
+      data.append(subData)
+    }
+    return data
+  }
+}
+
+public struct TypedData: ValueType {
+  public var types: Types
+  public var domain: Domain
+  public var primaryType: String
+  public var message: MessageObject
+
+  public static func value(from object: Any) throws -> TypedData {
+    guard let valueMaps = object as? [String: Any],
+      let typeDict: [String: Type] = try? valueMaps.value(for: "types") else {
+      throw EtherKitError.web3Failure(reason: .parsingFailure)
+    }
+    let types: Types = Types(types: typeDict)
+
+    guard let domain: Domain = try? valueMaps.value(for: "domain"),
+      let primaryType: String = try? valueMaps.value(for: "primaryType"),
+      let messageObject: [String: Any] = try? valueMaps.value(for: "message"),
+      let message = try? MessageObject(from: messageObject, typeName: primaryType, types: types) else {
+      throw EtherKitError.web3Failure(reason: .parsingFailure)
+    }
+
+    return TypedData(types: types, domain: domain, primaryType: primaryType, message: message)
+  }
+
+  public func getDomainSeparator() throws -> Data {
```

Comment:
perhaps with Hash in method name?

Reply:
Same thing like I said for `getHashStruct` I named this against the names used in the EIP.